### PR TITLE
Update tx struct

### LINKED ISSUE #115: Create better transaction representation
We need to find a good balance between ease of querying and keeping the transaction generic enough. 

The current transaction struct looks like this

```
type Tx struct {
	ID        string `json:"-" gorm:"primary_key;size:100;default:'6ba7b810-9dad-11d1-80b4-00c04fd430c8'"`
	To        uint64 `json:"to"`
	From      uint64 `json:"from"`
	Data      []byte `json:"data"`
	Signature []byte `json:"sig" gorm:"null"`
	TxHash    string `json:"hash" gorm:"not null"`
	Status    uint64 `json:"status"`
	Type      uint64 `json:"type" gorm:"not null"`
}
```

We need to query by fees, nonce, tokenType while picking transactions so the revised transaction could look something like below

```
type Tx struct {
	ID        string `json:"-" gorm:"primary_key;size:100;default:'6ba7b810-9dad-11d1-80b4-00c04fd430c8'"`
	
         To        uint64 `json:"to"`
	From     uint64 `json:"from"`
	Data      []byte `json:"data"`
         Nonce uint64 
         Fees uint64
         TokenType uint64
	
         Signature []byte `json:"sig" gorm:"null"`
         TxHash    string `json:"hash" gorm:"not null"`
	
        Status    uint64 `json:"status"`
	Type      uint64 `json:"type" gorm:"not null"`
}
```

### PR DESCRIPTION
**Pull request by vaibhavchellani**
Fixes #115

### CODE REVIEW

File: db/tx.go [hunk 1]
```diff
@@ -9,10 +9,12 @@ import (
 	"github.com/BOPR/wallet"
 	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/jinzhu/gorm"
 	"github.com/kilic/bn254/bls"
 )
 
 var (
+	// ErrBadSignature error bad signature
```
Comment (ChihChengLiang) [inline db/tx.go:17]: Can we avoid doc on self obvious code? The doc string does not provide helpful information.

File: db/tx.go [hunk 2]
```diff
@@ -100,12 +134,17 @@ func (DBI *DB) FetchTxType() (txType uint64, err error) {
...
+	}
+
 	// fetch from state MP
-	err = DBI.fetchMPWithID(tx.From, &fromMerkleProof)
+	err = DBI.fetchMPWithID(from, &fromMerkleProof)
 	if err != nil {
 		return
 	}
-	toState, err := DBI.GetStateByIndex(tx.To)
+	toState, err := DBI.GetStateByIndex(to)
```
Comment (ChihChengLiang) [inline db/tx.go:147]: So if the tx is a mass migration. `FetchFromAndToStateIDs` returns `to` stateID as 0, and we are fetching stateID 0 for no reason?

File: tests/tx_test.go [hunk 3]
```diff
@@ -1,91 +1,33 @@
...
-
-	txs, err := db.PopTxs()
-	if err != nil {
-		t.Errorf("PopTxs error %s", err)
-	}
-	for i, tx := range []core.Tx{tx2, tx3} {
-		assert.Equal(t, tx.TxHash, txs[i].TxHash)
-	}
-}
+// func setupDB(cfg config.Configuration) (DBI db.DB, cleanup func(), err error) {
```
Comment (ChihChengLiang) [inline tests/tx_test.go:7]: It's fine we just remove this test. We should avoid commit commentted off code to the codebase.

**Review by ChihChengLiang — APPROVED** [review]
Looks good after the commented issues are addressed

File: db/tx.go (hunk 1 above)
Comment (vaibhavchellani) [reply db/tx.go:17]: GoLang shows warnings :((

File: tests/tx_test.go (hunk 3 above)
Comment (vaibhavchellani) [reply tests/tx_test.go:7]: Will remove

File: db/tx.go (hunk 1 above)
Comment (ChihChengLiang) [reply db/tx.go:17]: I see, I'll look into how to disable that warning.

File: db/tx.go (hunk 2 above)
Comment (vaibhavchellani) [reply db/tx.go:147]: yeah good eye, created much simpler mass migrations witness fetcher.

File: db/tx.go [hunk 4]
```diff
@@ -149,6 +188,46 @@ func getWitnessTranfer(bz bazooka.Bazooka, DBI DB, tx core.Tx) (fromMerkleProof,
...
+
+	// apply the new from leaf
+	currentFromStateCopy := fromMerkleProof.State
+	currentFromStateCopy.Data = newFrom
+	err = dbCopy.UpdateState(currentFromStateCopy)
+	if err != nil {
+		return
+	}
+
+	return fromMerkleProof, toMerkleProof, dbCopy, nil
```
Comment (ChihChengLiang) [inline db/tx.go:228]: Where is this toMerkleProof come from? Or will it be just an empty data structure?

**Review by ChihChengLiang — APPROVED** [review]
Left a question. Otherwise looks good to me

File: db/tx.go (hunk 4 above)
Comment (vaibhavchellani) [reply db/tx.go:228]: It will be empty, just added so that it satisfies the interface, it isn't accessed anytime down the stack so should be okay

**Pull request merged**